File: frontend/protocol.js
/**
 * Binary protocol for e2hat encrypted chat.
 * JavaScript mirror of server/protocol.py — both must stay in sync.
 *
 * Frame format:
 *   [MSG_TYPE: 1 byte] [PAYLOAD_LEN: 2 bytes big-endian] [PAYLOAD: N bytes]
 *
 * Compressed elliptic curve point (secp521r1):
 *   [X coordinate: 66 bytes big-endian] [PARITY: 1 byte] = 67 bytes total
 *
 * All multi-byte integers are big-endian. Session IDs are 4 bytes.
 */

// secp521r1 coordinate = 66 bytes, compressed point = 67 bytes
const COORD_SIZE = 66;
const POINT_SIZE = 67;

// Handshake
const HELLO = 0x01;          // Client -> Server: here is my DH public key
const WELCOME = 0x02;        // Server -> Client: here is your MO public key

// Sender-side Massey-Omura 3-pass
const MO_SEND_INIT = 0x10;   // Client sends C1 = mo.encrypt(E) + destination
const MO_SEND_STEP2 = 0x11;  // Server returns C2 = server_mo.encrypt(C1)
const MO_SEND_STEP3 = 0x12;  // Client sends C3 = mo.decrypt(C2)

// Receiver-side Massey-Omura 3-pass
const MO_RECV_INIT = 0x20;   // Server sends C1' = server_mo.encrypt(E) to receiver
const MO_RECV_STEP2 = 0x21;  // Receiver returns C2' = mo.encrypt(C1')
const MO_RECV_STEP3 = 0x22;  // Server sends C3' = server_mo.decrypt(C2') to receiver

// Control
const ERROR = 0x30;           // Error with 1-byte error code
const PEER_ONLINE = 0x31;     // A peer connected
const PEER_OFFLINE = 0x32;    // A peer disconnected
const MSG_ACK = 0x33;         // Server received the message (sender MO complete)
const MSG_DELIVERED = 0x34;   // Message delivered to recipient
const MSG_QUEUED = 0x35;      // Recipient offline, message queued

// Error codes
const ERR_USER_NOT_FOUND = 0x01;
const ERR_INVALID_FRAME = 0x02;
const ERR_SESSION_EXPIRED = 0x03;

/**
 * Convert a BigInt to a fixed-size big-endian Uint8Array.
 */
function bigintToBytes(value, length) {
    const bytes = new Uint8Array(length);
    let v = value;
    for (let i = length - 1; i >= 0; i--) {
        bytes[i] = Number(v & 0xFFn);
        v >>= 8n;
    }
    return bytes;
}

/**
 * Convert a Uint8Array (big-endian) to BigInt.
 */
function bytesToBigint(bytes, offset, length) {
    let result = 0n;
    for (let i = 0; i < length; i++) {
        result = (result << 8n) | BigInt(bytes[offset + i]);
    }
    return result;
}

function packPoint(x, parity) {
    const buf = new Uint8Array(POINT_SIZE);
    buf.set(bigintToBytes(x, COORD_SIZE), 0);
    buf[COORD_SIZE] = Number(parity);
    return buf;
}

function unpackPoint(data, offset = 0) {
    const x = bytesToBigint(data, offset, COORD_SIZE);
    const parity = BigInt(data[offset + COORD_SIZE]);
    return { x, parity, nextOffset: offset + POINT_SIZE };
}

function packFrame(msgType, payload) {
    const frame = new Uint8Array(3 + payload.length);
    frame[0] = msgType;
    frame[1] = (payload.length >> 8) & 0xFF;
    frame[2] = payload.length & 0xFF;
    frame.set(payload, 3);
    return frame;
}

function unpackFrame(data) {
    const arr = new Uint8Array(data);
    const msgType = arr[0];
    const payloadLen = (arr[1] << 8) | arr[2];
    const payload = arr.slice(3, 3 + payloadLen);
    return { msgType, payload };
}

// --- Pack functions ---

function packHello(pubkeyX, pubkeyParity) {
    return packFrame(HELLO, packPoint(pubkeyX, pubkeyParity));
}

function packMoSendInit(destX, destParity, j, c1X, c1Parity) {
    const dest = packPoint(destX, destParity);
    const c1 = packPoint(c1X, c1Parity);
    const payload = new Uint8Array(POINT_SIZE + 1 + POINT_SIZE);
    payload.set(dest, 0);
    payload[POINT_SIZE] = j;
    payload.set(c1, POINT_SIZE + 1);
    return packFrame(MO_SEND_INIT, payload);
}

function packMoSendStep3(sessionId, c3X, c3Parity) {
    const point = packPoint(c3X, c3Parity);
    const payload = new Uint8Array(4 + POINT_SIZE);
    const dv = new DataView(payload.buffer);
    dv.setUint32(0, sessionId);
    payload.set(point, 4);
    return packFrame(MO_SEND_STEP3, payload);
}

function packMoRecvStep2(sessionId, c2X, c2Parity) {
    const point = packPoint(c2X, c2Parity);
    const payload = new Uint8Array(4 + POINT_SIZE);
    const dv = new DataView(payload.buffer);
    dv.setUint32(0, sessionId);
    payload.set(point, 4);
    return packFrame(MO_RECV_STEP2, payload);
}

// --- Unpack functions ---

function unpackWelcome(payload) {
    return unpackPoint(payload);
}

function unpackMoSendStep2(payload) {
    const dv = new DataView(payload.buffer, payload.byteOffset, payload.byteLength);
    const sessionId = dv.getUint32(0);
    const { x, parity } = unpackPoint(payload, 4);
    return { sessionId, x, parity };
}

function unpackMoRecvInit(payload) {
    const dv = new DataView(payload.buffer, payload.byteOffset, payload.byteLength);
    const sessionId = dv.getUint32(0);
    let off = 4;
    const sender = unpackPoint(payload, off);
    off = sender.nextOffset;
    const j = payload[off];
    off += 1;
    const c1 = unpackPoint(payload, off);
    return {
        sessionId,
        senderX: sender.x,
        senderParity: sender.parity,
        j,
        c1X: c1.x,
        c1Parity: c1.parity,
    };
}

function unpackMoRecvStep3(payload) {
    const dv = new DataView(payload.buffer, payload.byteOffset, payload.byteLength);
    const sessionId = dv.getUint32(0);
    const { x, parity } = unpackPoint(payload, 4);
    return { sessionId, x, parity };
}

function unpackError(payload) {
    return payload[0];
}

function unpackPeerEvent(payload) {
    return unpackPoint(payload);
}

function unpackSessionId(payload) {
    const dv = new DataView(payload.buffer, payload.byteOffset, payload.byteLength);
    return dv.getUint32(0);
}

window.Protocol = {
    HELLO, WELCOME,
    MO_SEND_INIT, MO_SEND_STEP2, MO_SEND_STEP3,
    MO_RECV_INIT, MO_RECV_STEP2, MO_RECV_STEP3,
    ERROR, PEER_ONLINE, PEER_OFFLINE,
    MSG_ACK, MSG_DELIVERED, MSG_QUEUED,
    ERR_USER_NOT_FOUND, ERR_INVALID_FRAME, ERR_SESSION_EXPIRED,
    COORD_SIZE, POINT_SIZE,
    bigintToBytes, bytesToBigint,
    packPoint, unpackPoint,
    packFrame, unpackFrame,
    packHello, packMoSendInit, packMoSendStep3, packMoRecvStep2,
    unpackWelcome, unpackMoSendStep2, unpackMoRecvInit, unpackMoRecvStep3,
    unpackError, unpackPeerEvent, unpackSessionId,
};

File: server/protocol.py
"""Binary protocol for e2hat encrypted chat.

Every message exchanged over the WebSocket follows this frame format:

    [MSG_TYPE: 1 byte] [PAYLOAD_LEN: 2 bytes big-endian] [PAYLOAD: N bytes]

Elliptic curve points (secp521r1) are compressed to 67 bytes:

    [X coordinate: 66 bytes big-endian] [PARITY: 1 byte (0 or 1)]

Protocol phases:
    0x01-0x03  Handshake     HELLO → WELCOME (MO key + challenge) → AUTH (ECDSA sig)
    0x10-0x12  Send (MO)     Sender ↔ Server Massey-Omura 3-pass with signature relay
    0x20-0x22  Receive (MO)  Server ↔ Receiver Massey-Omura 3-pass with signature relay
    0x30-0x35  Control       Errors, presence, delivery status (ack/delivered/queued)
"""

import struct

# secp521r1 coordinate is 66 bytes (521 bits, rounded up to 66 bytes)
COORD_SIZE = 66
# Compressed point = X coordinate + 1 byte parity
POINT_SIZE = 67

# --- Message types ---
# Handshake: client sends HELLO with its DH public key, server replies WELCOME
# with a per-client Massey-Omura public key and a 32-byte challenge.
# Client responds with AUTH containing the signature of the challenge.
HELLO = 0x01
WELCOME = 0x02
AUTH = 0x03

# Sender-side Massey-Omura 3-pass:
#   SEND_INIT:  client sends encrypted point + destination
#   SEND_STEP2: server double-encrypts and returns to client
#   SEND_STEP3: client removes its layer, server can now decrypt to get E
MO_SEND_INIT = 0x10
MO_SEND_STEP2 = 0x11
MO_SEND_STEP3 = 0x12

# Receiver-side Massey-Omura 3-pass:
#   RECV_INIT:  server sends double-encrypted point to receiver
#   RECV_STEP2: receiver adds its layer and returns
#   RECV_STEP3: server removes its layer, receiver can now decrypt to get E
MO_RECV_INIT = 0x20
MO_RECV_STEP2 = 0x21
MO_RECV_STEP3 = 0x22

# Control messages
ERROR = 0x30         # Server reports an error condition
PEER_ONLINE = 0x31   # A peer connected to this server
PEER_OFFLINE = 0x32  # A peer disconnected from this server
MSG_ACK = 0x33       # Server received the message from sender (MO complete)
MSG_DELIVERED = 0x34  # Message was delivered to the recipient
MSG_QUEUED = 0x35    # Recipient is offline, message queued for later delivery


def pack_point(x, parity):
    """Pack a compressed point into 67 bytes."""
    return x.to_bytes(COORD_SIZE, "big") + struct.pack("B", parity)


def unpack_point(data, offset=0):
    """Unpack a compressed point from bytes. Returns (x, parity, new_offset)."""
    if len(data) < offset + POINT_SIZE:
        raise ValueError(f"Payload too short for point at offset {offset}: need {offset + POINT_SIZE}, got {len(data)}")
    x = int.from_bytes(data[offset:offset + COORD_SIZE], "big")
    parity = data[offset + COORD_SIZE]
    return x, parity, offset + POINT_SIZE


def pack_frame(msg_type, payload):
    """Pack a protocol frame."""
    return struct.pack("B", msg_type) + struct.pack(">H", len(payload)) + payload


def unpack_frame(data):
    """Unpack a protocol frame. Returns (msg_type, payload)."""
    if len(data) < 3:
        raise ValueError(f"Frame too short: need at least 3 bytes, got {len(data)}")
    msg_type = data[0]
    payload_len = struct.unpack(">H", data[1:3])[0]
    if len(data) < 3 + payload_len:
        raise ValueError(f"Frame truncated: header says {payload_len} bytes, got {len(data) - 3}")
    payload = data[3:3 + payload_len]
    return msg_type, payload


# --- Pack functions ---
# Each pack function builds a complete frame (header + payload) ready to send.

def pack_hello(client_pubkey_x, client_pubkey_parity):
    payload = pack_point(client_pubkey_x, client_pubkey_parity)
    return pack_frame(HELLO, payload)


def pack_welcome(server_mo_pubkey_x, server_mo_pubkey_parity, challenge):
    payload = pack_point(server_mo_pubkey_x, server_mo_pubkey_parity)
    payload += challenge  # 32 bytes
    return pack_frame(WELCOME, payload)


def unpack_welcome(payload):
    x, parity, off = unpack_point(payload)
    challenge = payload[off:off + 32]
    return x, parity, challenge


def pack_auth(r, s):
    """Pack ECDSA signature (r, s scalars)."""
    return pack_frame(AUTH, r.to_bytes(COORD_SIZE, "big") + s.to_bytes(COORD_SIZE, "big"))


def unpack_auth(payload):
    """Unpack ECDSA signature."""
    if len(payload) < COORD_SIZE * 2:
        raise ValueError(f"AUTH payload too short: need {COORD_SIZE * 2}, got {len(payload)}")
    r = int.from_bytes(payload[:COORD_SIZE], "big")
    s = int.from_bytes(payload[COORD_SIZE:COORD_SIZE*2], "big")
    return r, s


def pack_mo_send_init(dest_pubkey_x, dest_pubkey_parity, j, c1_x, c1_parity, sig_r=None, sig_s=None, timestamp_ms=None):
    payload = pack_point(dest_pubkey_x, dest_pubkey_parity)
    payload += struct.pack("B", j)
    payload += pack_point(c1_x, c1_parity)
    if sig_r is not None and sig_s is not None:
        payload += sig_r.to_bytes(COORD_SIZE, "big") + sig_s.to_bytes(COORD_SIZE, "big")
        payload += struct.pack(">Q", timestamp_ms or 0)
    return pack_frame(MO_SEND_INIT, payload)


def pack_mo_send_step2(session_id, c2_x, c2_parity):
    payload = struct.pack(">I", session_id) + pack_point(c2_x, c2_parity)
    return pack_frame(MO_SEND_STEP2, payload)


def pack_mo_send_step3(session_id, c3_x, c3_parity):
    payload = struct.pack(">I", session_id) + pack_point(c3_x, c3_parity)
    return pack_frame(MO_SEND_STEP3, payload)


def pack_mo_recv_init(
    session_id, sender_pubkey_x, sender_pubkey_parity, j,
    c1_x, c1_parity, sig_r=None, sig_s=None, timestamp_ms=None,
):
    payload = struct.pack(">I", session_id)
    payload += pack_point(sender_pubkey_x, sender_pubkey_parity)
    payload += struct.pack("B", j)
    payload += pack_point(c1_x, c1_parity)
    if sig_r is not None and sig_s is not None:
        payload += sig_r.to_bytes(COORD_SIZE, "big") + sig_s.to_bytes(COORD_SIZE, "big")
        payload += struct.pack(">Q", timestamp_ms or 0)
    return pack_frame(MO_RECV_INIT, payload)


def pack_mo_recv_step2(session_id, c2_x, c2_parity):
    payload = struct.pack(">I", session_id) + pack_point(c2_x, c2_parity)
    return pack_frame(MO_RECV_STEP2, payload)


def pack_mo_recv_step3(session_id, c3_x, c3_parity):
    payload = struct.pack(">I", session_id) + pack_point(c3_x, c3_parity)
    return pack_frame(MO_RECV_STEP3, payload)


def pack_error(error_code):
    return pack_frame(ERROR, struct.pack("B", error_code))


def pack_peer_online(peer_pubkey_x, peer_pubkey_parity):
    payload = pack_point(peer_pubkey_x, peer_pubkey_parity)
    return pack_frame(PEER_ONLINE, payload)


def pack_peer_offline(peer_pubkey_x, peer_pubkey_parity):
    payload = pack_point(peer_pubkey_x, peer_pubkey_parity)
    return pack_frame(PEER_OFFLINE, payload)


# --- Unpack functions ---
# Each unpack function receives a payload (without the frame header) and returns
# the parsed fields. The frame header is stripped by unpack_frame() first.

def unpack_hello(payload):
    x, parity, _ = unpack_point(payload)
    return x, parity


def unpack_mo_send_init(payload):
    dest_x, dest_parity, off = unpack_point(payload)
    if len(payload) < off + 1:
        raise ValueError("MO_SEND_INIT payload too short for j byte")
    j = payload[off]
    off += 1
    c1_x, c1_parity, off2 = unpack_point(payload, off)
    sig_r, sig_s, timestamp_ms = None, None, None
    if len(payload) >= off2 + COORD_SIZE * 2 + 8:
        sig_r = int.from_bytes(payload[off2:off2 + COORD_SIZE], "big")
        sig_s = int.from_bytes(payload[off2 + COORD_SIZE:off2 + COORD_SIZE * 2], "big")
        timestamp_ms = struct.unpack(">Q", payload[off2 + COORD_SIZE * 2:off2 + COORD_SIZE * 2 + 8])[0]
    return dest_x, dest_parity, j, c1_x, c1_parity, sig_r, sig_s, timestamp_ms


def unpack_mo_send_step3(payload):
    if len(payload) < 4:
        raise ValueError("MO_SEND_STEP3 payload too short for session ID")
    session_id = struct.unpack(">I", payload[:4])[0]
    c3_x, c3_parity, _ = unpack_point(payload, 4)
    return session_id, c3_x, c3_parity


def unpack_mo_recv_step2(payload):
    if len(payload) < 4:
        raise ValueError("MO_RECV_STEP2 payload too short for session ID")
    session_id = struct.unpack(">I", payload[:4])[0]
    c2_x, c2_parity, _ = unpack_point(payload, 4)
    return session_id, c2_x, c2_parity


def pack_msg_ack(session_id):
    return pack_frame(MSG_ACK, struct.pack(">I", session_id))


def pack_msg_delivered(session_id):
    return pack_frame(MSG_DELIVERED, struct.pack(">I", session_id))


def pack_msg_queued(session_id):
    return pack_frame(MSG_QUEUED, struct.pack(">I", session_id))


# Error codes sent inside ERROR frames
ERR_USER_NOT_FOUND = 0x01   # Destination public key not connected to this server
ERR_INVALID_FRAME = 0x02    # Malformed or unrecognized frame
ERR_SESSION_EXPIRED = 0x03  # Session ID not found or already completed
ERR_RATE_LIMITED = 0x04     # Too many messages, slow down
